implement global window.onerror handler

sasjonbri opened the pull request:
Add globalErrorHandler before other JavaScript
content is loaded to catch any error.

A globalErrorHandler.js file is loaded by the grunt build
in concat.js which attaches a listener to window.onerror

If unchecked error occurs during bootup or when
the framework is not available, render a vanillaJS
error dialog. Handle i18n and css inline to keep
self-contained.

Otherwise, render an implementation of sap.m.Popup
with a textarea showing technical details of the error.

akudev commented on src/sap.ui.commons/src/sap/ui/commons/themes/base/globalErrorHandler.less:
```diff
@@ -0,0 +1,24 @@
+.sapMMsgBoxGEH .expandableTextArea {
```
Even when prefixed with a UI5-specific CSS class within this selector, you should not introduce an unprefixed CSS class name like "expandableTextArea": the place where you write it to the HTML is at risk of being styled wrongly by CSS which is added by other libraries or the application.

akudev commented on src/sap.ui.commons/src/sap/ui/commons/themes/base/globalErrorHandler.less:
```diff
@@ -0,0 +1,24 @@
+	width: 100%;
+	height: 100%;
+	font-size: 12px;
+	resize:none;
```
nit: consistent spacing (space after "resize:") would be nice

akudev commented on src/sap.ui.core/src/sap/ui/core/Configuration.js:
```diff
@@ -786,7 +787,7 @@ sap.ui.define(['jquery.sap.global', 'sap/ui/base/Object', './Locale', 'sap/ui/th
 
 		/**
-		 * Return whether the controller code is deactivated. During design mode the 
+		 * Return whether the controller code is deactivated. During design mode the
```
Huh, sentence abruptly ends - but that's not your mistake. I told the developer who broke this.

akudev commented on src/sap.ui.commons/src/sap/ui/commons/themes/base/globalErrorHandler.less:
```diff
@@ -0,0 +1,24 @@
+	height: 200px;
+	padding: 10px;
+}
+.sapMMsgBoxGEH .expandableTextArea textarea {
```
For performance reasons, please try to avoid selectors that end with a plain HTML tag name and have no "direct child" operator before that tag name. There will not be too many "textarea" tags in the page, so this instance is not really an issue, and only a star ("*") in the end would be really bad, but it's good practice to use a ">" between parts of a selector if possible.
Not sure... is this a direct child in the end? If the structure is more complex, you might just ignore this comment or add a class name to the HTML tag in critical cases.

akudev commented:
Thanks for the proposal so far!
The code owner also needs to have a look. This will happen next, as discussed by direct mail exchange.

sasjonbri commented:
Thanks for the feedback Andreas.

I changed the css class I was adding to sapMMsgBoxGEHExpTextArea. Hopefully that seems like an appropriate prefixed name.
I also added a direct descendant selector to the textarea style.

I rebased and squashed my changes against master...let me know if that's the pattern you guys prefer, or if you would like the commits that address comments to be left separate.

akudev commented:
"sapMMessageBoxGEHExpTextArea" - lol, that must be the longest CSS class we have, or at least the one squeezing in as many (abbreviated) words as possible... but thanks for changing.

I think squashing is fine. Certainly for this patch when the code owner newly looks at it now.
Internally we use Gerrit for code reviews, so we have one commit, but still see the diff between different patchsets...

akudev commented:
Hi Jonathan,

as discussed between you and Frederic in person and later by mail, we are awaiting your updated pull request. No pressure, just documenting it because nothing happened for a long time. :-)

Regards
Andreas

jonbri commented:
Hello,

I uploaded a re-vamped version of the global error handler support.

Changes:
- Rather than using MessageBox, I created a small control: sap.m.ErrorDialog
- Consumers can "opt-in" using the experimental property: xx-enableglobalerrorhandler
- Simplified implementation
- hook-points pre and post dialog invocation

I look forward to any feedback,
Jonathan

Michadelic commented:
Hello Andreas,

Jonathan uploaded a new commit, can you check again?
Was there already an agreement to merge it?

Kind Regards,
Michael

akudev commented:
@Michadelic: it's the product owner's task to look at this and decide how to proceed, e.g. to assign time for a developer to look into it. He is informed and was also in direct communication with Jonathan before, but as you see, there is no feedback yet.

frdrcbrg commented:
Hi Jonathan, we have taken the review of your pull request into our current sprint. We will update the contribution with feedback if required in the next 3 weeks.

Thanks!
Frederic

sasjonbri commented:
Closing as code is out-dated. If there is still interest in this pull-request I can update the code and re-open
